fuzz/conversion.cpp

```diff
@@ -0,0 +1,707 @@
+      std::abort();
+    }
+    os << "ASSERT_EQUAL(outlen, 1234);\n";
+    os << "std::vector<" << nameoftype(ToType{}) << "> output(outlen);\n";
+    os << "const auto r = implementation." << name << "((const "
+       << nameoftype(FromType{}) << "*) data\n, data_len\n, output.data());\n";
+
+    if constexpr (std::is_same_v<ConversionResult, simdutf::result>) {
+      os << " ASSERT_EQUAL(r.error,simdutf::error_code::SUCCESS);\n";
+      os << " ASSERT_EQUAL(r.count,1234);\n";
+    } else {
+      os << "   ASSERT_EQUAL(r, 1234);\n";
+    }
+
+    // dump the output data
+    os << "const std::vector<" << nameoftype(ToType{}) << "> expected_out{};\n";
+    os << " ASSERT_TRUE(output.size()==expected_out.size());\n";
+    os << " for(std::size_t i=0; i<output.size(); ++i) { "
+          "ASSERT_EQUAL(+output.at(i),+expected_out.at(i));};\n";
+
+    os << "}\n";
+    os << "// end testcase\n";
+  }
+};
+
+const auto populate_functions() {
+  // build a tuple containing function pointers to everything we want to fuzz
+  using I = simdutf::implementation;
+
+#define ADD(lenfunc, conversionfunc)                                           \
+  Conversion<ENCODING_FROM_CONVERSION_NAME(&I::conversionfunc),                \
+             ENCODING_TO_CONVERSION_NAME(&I::conversionfunc),                  \
+             decltype(&I::lenfunc), decltype(&I::conversionfunc)> {            \
+    &I::lenfunc, &I::conversionfunc, std::string{NAMEOF(&I::lenfunc)},         \
+        std::string {                                                          \
+      NAMEOF(&I::conversionfunc)                                               \
+    }                                                                          \
+  }
+
+// useful during development to disable fuzzers that found something so fuzzing
+// can continue on the others, without making the corpus useless because the
+// action taken from the first element means something else when the elements
+// shift around
+#define IGNORE(len, x)                                                         \
+  Conversion<ENCODING_FROM_CONVERSION_NAME(&I::x),                             \
+             ENCODING_TO_CONVERSION_NAME(&I::x), decltype(&I::len),            \
+             decltype(&I::x)> {                                                \
+    nullptr, nullptr, std::string{NAMEOF(&I::len)}, std::string {              \
+      NAMEOF(&I::x)                                                            \
+    }                                                                          \
+  }
+  std::tuple cases_requiring_valid_data(
+      // all these cases require valid input for invoking the convert function
+      ADD(latin1_length_from_utf16, convert_valid_utf16be_to_latin1),
+      ADD(utf32_length_from_utf16be, convert_valid_utf16be_to_utf32),
+      ADD(utf8_length_from_utf16be, convert_valid_utf16be_to_utf8),
+
+      ADD(latin1_length_from_utf16, convert_valid_utf16le_to_latin1),
```

Comment:
`convert_valid_utf16be_to_latin1` is not part of our public API (we don't document it in the README) and it requires that the input is valid and can be converted to latin1.